### ISSUE TITLE
Implement support for generic challenge-response style authentication

### ISSUE THREAD

**Issue by conradgrobler** [project-oak/oak#1357]
At the moment we use bearer tokens to do pseudonymous authentication. It would be better not to pass secret tokens or keys to the server to do authentication, seeing that a leaked secret allows impersonation. Challenge-response protocols allow for authentication without sending secrets to the server, but cannot be implemented by just attaching metadata to each call due to the interactive nature.

Mutually authenticated TLS connections with client certificates provide an option for authenticating clients, but it is somewhat inconvenient and requires a root of trust for the client certificates (typicall the subject of the certificate is seen as the identity, rather than using the public key as an identifier).

A more generic option is to provide a gRPC-based challenge response authentication mechanism that allows a client to 'upgrade' an existing connection with an identity after the TLS connection has been established.

The idea is roughly

- The client establishes a TLS connection to the server
- The client asks for a challenge via a gRPC call
  - In the OpenID Connect case this includes the OIDC server details
  - For a pseudonymous signature-based approach this will just be a random byte string
- The client produces a response and sends it to the server via another gRPC call to attach an identity to the current connection
  - For OIDC this is the ID token
  - For a pseudonymous signature-based approach the client will sign the challenge and send the signature and public key. The public key will act as an identifier for the user if the signature is validated.
- All future calls on the connection will then have the privilege associated with the assoicated identity
- It should be possible for a client to attach multiple identities to a single connection by repeating the process

A prerequisite for this is #1020 (supporting per-connection state on the gRPC server node), which has been on hold for a while. The server must be able to associate the challenge sent to the client and the received identity with the current connection.

**Comment by tiziano88** [comment]
An alternative approach may be for the client to just sign the content of each outgoing request (including headers and body).

Advantages:

- no need to keep state on the server
- can sign each request differently if necessary
- provides non-repudiation for requests (server can store the payload and signature as proof of the request actually coming from the client with that public key)

Disadvantages:

- vulnerable to replay attacks (though only by the other endpoint of the TLS connection, so probably not a big deal)
- more logic on the client
- may not work with streaming requests

**Comment by tiziano88** [comment]
BTW in any case let's make sure that we use the same signing / verification logic that @ipetr0v is using for the wasm module signatures

**Comment by ipetr0v** [comment]
We also can implement [Kerberos](https://tools.ietf.org/html/rfc4120) like client authentication (token distribution server). And this server could be a part of Oak Runtime, just on different port.
And we can make it pseudonymous using [pkinit](https://tools.ietf.org/html/rfc4556), i.e. users will not need to enter any passwords, and use public keys instead.

Also after talking to @conradgrobler - authentication that relies on temporary tokens will require to have trusted time (non-global one).

**Comment by ipetr0v** [comment]
Hm, looks like Kerberos pkinit also requires that client's key is signed by the CA:
```
   2. The KDC tests the client's request against its authentication
      policy and trusted Certification Authorities (CAs).
```
where KDC is a Key Distribution Service

**Comment by ipetr0v** [comment]
About the option of encrypting all requests with a private key and sending a corresponding public key in the Metadata:
`tonic` has an [`Interceptor`](https://docs.rs/tonic/0.2.1/tonic/struct.Interceptor.html) but it cannot be used to change the `Message`.
```
These interceptors do not allow you to modify the Message of the request but allow you to check for metadata.
If you would like to apply middleware like features to the body of the request, going through the tower abstraction is recommended.
```

**Comment by tiziano88** [comment]
- the point is to **sign** the message, not to encrypt it
- > gRPC interceptors are similar to middleware but have much less flexibility. This interceptor allows you to do two main things, one is to add/remove/check items in the MetadataMap of each request. Two, cancel a request with any Status.
  + so it seems fine for us? it needs to sign the body + existing metadata, and add the result as a new metadata field (+ the public key too, possibly)

**Comment by conradgrobler** [comment]
Looking at the code, it seems that the message is not available to the inceptor:
https://docs.rs/tonic/0.3.0/src/tonic/interceptor.rs.html#33

**Comment by tiziano88** [comment]
> Looking at the code, it seems that the message is not available to the interceptor:
> https://docs.rs/tonic/0.3.0/src/tonic/interceptor.rs.html#33

good find, I didn't notice that, it seems that the interceptor intentionally discards the message body?

**Comment by tiziano88** [comment]
would it be possible to intercept things in `tower` instead? https://docs.rs/tower/0.3.1/tower/trait.Service.html

**Comment by ipetr0v** [comment]
Possibly, but I think we will loose the ability to use `tonic` abstractions.

**Comment by tiziano88** [comment]
I'm still hoping at some point we can actually merge gRPC and HTTP server pseudo node implementation, and I guess if / when that happens they will just rely on `tonic` anyways.

Would the proposed auth solution work for the HTTP pseudo node?

cc @rbehjati

**Comment by conradgrobler** [comment]
> Possibly, but I think we will loose the ability to use `tonic` abstractions.

I think it is possible to just add a middleware service and still use the `tonic` abstractions. Currently we create a channel (which implements `GrpcService<BoxBody>`) and pass that as the `inner` parameter to `with_interceptor` when creating the client. If we create a new `GrpcService<BoxBody>` that wraps the channel and handles the signing of the content, we can pass that as the `inner` parameter to `with_interceptor`.

**Comment by ipetr0v** [comment]
Just to confirm, `tonic` specifically doesn't enable client authentication if client's root CA is not specified:
https://github.com/hyperium/tonic/blob/04a8c0c82a4007f48c3bf3539a3f2312746fedd1/tonic/src/transport/service/tls.rs#L130-L131

```rust
        let mut config = match client_ca_root {
            None => ServerConfig::new(NoClientAuth::new()),
```

**Comment by ipetr0v** [comment]
Trying to add an intermediate gRPC service to implement message signing and found an interesting `tonic` feature:
`GrpcService<BoxBody>::call(...)` works with `http::Request` and not `tonic::Request`, and thus gRPC `metadata` is not available in this function.

**Comment by tiziano88** [comment]
BTW @conradgrobler is this proposed challenge-response user auth scheme vulnerable to the same relay / proxy attack described in #1374 ?

**Comment by conradgrobler** [comment]
I guess it is. A malicious application operator can terminate the TLS in a proxy to obtain a MitM position. Once the user has elevated the privileges of the connection via authentication, the operator can then re-use the existing connection to perform arbitrary gRPC calls as the user.

**Comment by rbehjati** [comment]
> I'm still hoping at some point we can actually merge gRPC and HTTP server pseudo node implementation, and I guess if / when that happens they will just rely on `tonic` anyways.
> 
> Would the proposed auth solution work for the HTTP pseudo node?
> 
> cc @rbehjati

I think it should be possible to have a common way of creating gRPC and HTTP server pseudo-nodes. I have created #1390 to track that. In my implementation of TLS support for HTTP server pseudo-node (#1370 ), I have used `tokio-rustls` and `hyper-rustls`. I feel like for implementing the challenge-response described here, we have to use even lower level crates.

I don't think `tonic` can be used for creating an HTTP server. So, I am not sure if it is a good idea to try and implement the ideas discussed here using `tonic` (e.g., using `interceptors`), if we want to merge the implementations of gRPC and HTTP server pseudo-nodes.

**Comment by tiziano88** [comment]
@conradgrobler @rbehjati would it be possible to implement this (or something similar to it) for the HTTP server pseudo-node already? Or is the per-connection state still an issue there too? we can figure out how to do it for gRPC later on, in the meanwhile we may learn something about how the rest of the auth protocol is supposed to work

**Comment by conradgrobler** [comment]
The per-connection state is not yet implemented on the HTTP server pseudo-node, but it should be significantly simpler to add it there, seeing that it already supports lower-level access to the underlying connection acceptor.

**Comment by tiziano88** [comment]
As a temporary solution, and as improvement over the current bearer token mechanism, I suggest the following:

- Oak Runtime generates a single "challenge" at startup, and stores it somewhere internally
- exposes an endpoint (HTTP or gRPC) providing access to the challenge to clients
- client gets challenge, then signs is with private key (EDDSA or similar, as used elsewhere in Oak)
- this signed challenge is provided as part of each request to the server

This is similar to a bearer token, but it is scoped per runtime instance, so it seems a net improvement over the current mechanism.

@conradgrobler @rbehjati  WDYT?

**Comment by conradgrobler** [comment]
I agree that it is an incremental improvement over the bearer tokens. While it does not stop replay attacks, it limits the scope of the replay to a single runtime instance.

The one potential problem I can foresee is that it might cause problems when using a load-balancing solution, or in future when we implement a distributed runtime.

To support load balancing, it might be useful to provide a mechanism to specify a seed as a start-up parameter that can be used to create a consistent challenge across multiple instance if needed (e.g. if a seed is specified, generate a hash of the seed to produce the challenge).

**Comment by tiziano88** [comment]
I suggest making this change incrementally:

- first move from the current bearer token (+HMAC) mechanism, to using a signature over a fixed string (e.g. "oak-auth") to prove the possession of a private key (which is prone to replay attacks, of course)
- then make the string dynamic (i.e. the challenge) so that the signature is only bound to the specific runtime instance and cannot be replayed to other instances

**Comment by ipetr0v** [comment]
> first move from the current bearer token (+HMAC) mechanism, to using a signature over a fixed string (e.g. "oak-auth") to prove the possession of a private key (which is prone to replay attacks, of course)

As a first implementation iteration - should this string be public (and clients need to encrypt it in advance), or it should be a part of the challenge-response?

**Comment by tiziano88** [comment]
There is no challenge response now.

The string should be public and well known, basically equivalent to the current logic in

https://github.com/project-oak/oak/blob/d0eacef9642e2c2e85610a0331aedc50773a94c4/oak/client/application_client.h#L48

**Comment by tiziano88** [comment]
For the runtime-level challenge generation, I suppose we could just expose it over gRPC, in the same way we were considering for the connection-level challenge: a special gRPC method would offer the challenge, which would be returned by the runtime; the challenge will be the same for each connection to the same instance of the runtime, but there are no issues about load balancing, since the connection is already established.

**Comment by rbehjati** [comment]
> As a temporary solution, and as improvement over the current bearer token mechanism, I suggest the following:
> 
> * Oak Runtime generates a single "challenge" at startup, and stores it somewhere internally
> * exposes an endpoint (HTTP or gRPC) providing access to the challenge to clients
> * client gets challenge, then signs is with private key (EDDSA or similar, as used elsewhere in Oak)
> * this signed challenge is provided as part of each request to the server
> 
> This is similar to a bearer token, but it is scoped per runtime instance, so it seems a net improvement over the current mechanism.
> 
> @conradgrobler @rbehjati WDYT?

I am not sure I am following. Why do we need a temporary solution? Is it because we don't currently have a per-connection state? As Conrad mentioned it might not be too difficult to implement that for HTTP server node. I don't know what pre-connection state may need to entail (I guess the challenge and the public key), but I could look into that after I am done with my struggle with TLS streams.

**Comment by tiziano88** [comment]
Yes the temporary solution is while we figure out how to do per-connection state.

My suggestion is to do it in the following order:

- pre-determined fixed "challenge" (well known string -- not a real challenge)
- per-runtime challenge (should be easy enough to do via a `RwLock`, or even without any synchronization, if properly initialized)
- per-connection challenge (needs more work to figure how to do it for both gRPC and HTTP)

**Comment by rbehjati** [comment]
To be more specific, is this what you are suggesting @tiziano88 ?

> - pre-determined fixed "challenge" (well known string -- not a real challenge)
- The client establishes a TLS connection to the server
- The client asks for a challenge via a gRPC/HTTP call
- The server sends the well known string (e.g., "challenge")
- The client produces a response (signed challenge and public key) and sends it to the server via another gRPC/HTTP call
- All future calls on the connection will then have the privilege associated with the associated identity (this is from Conrad's description above, and I don't quite understand it. Can you explain @conradgrobler? Does the client have to sign all the messages or include the signature for the challenge in a header or something? What does the server do with the signature and public key that it receives from the client?)

> - per-runtime challenge (should be easy enough to do via a RwLock, or even without any synchronization, if properly initialized)
  - The only difference between this and the previous case is that each runtime has its own challenge word, but use that for all incoming connections. Is that right? 
> - per-connection challenge (needs more work to figure how to do it for both gRPC and HTTP)
- Each connection gets a different challenge. The same user may get different challenges for different connections. Is it correct?
 
From the discussions it sounds like "per-runtime challenge" would cause the load-balancing problem, and the "per-connection challenge" approach can solve it. Does the "per-connection challenge"  inherently solve this problem?! Or addressing this issue is what you mean by the "more work" that is needed for "per-connection challenge" @tiziano88 ?

**Comment by tiziano88** [comment]
> To be more specific, is this what you are suggesting @tiziano88 ?
> 
> > * pre-determined fixed "challenge" (well known string -- not a real challenge)
> 
> * The client establishes a TLS connection to the server
> * The client asks for a challenge via a gRPC/HTTP call
> * The server sends the well known string (e.g., "challenge")

In the well-known string case, there is nothing for the client to fetch, the string is literally a constant and well known by the client already, e.g. literal "oak-challenge". Very similar to what is already implemented for HMAC-based token authentication:

https://github.com/project-oak/oak/blob/d0eacef9642e2c2e85610a0331aedc50773a94c4/oak/client/application_client.h#L33

> * The client produces a response (signed challenge and public key) and sends it to the server via another gRPC/HTTP call
> * All future calls on the connection will then have the privilege associated with the associated identity (this is from Conrad's description above, and I don't quite understand it. Can you explain @conradgrobler? Does the client have to sign all the messages or include the signature for the challenge in a header or something? What does the server do with the signature and public key that it receives from the client?)

I think both alternatives are possible:

- a one-off special message with the signed challenge, which "upgrades" the entire connection from that point on
- just attach the signed challenge as metadata to each message

Perhaps the first one is more in line with the next steps though (and conceptually it makes sense to provide an identity proof only once), so perhaps we should aim for that directly.

> 
> > * per-runtime challenge (should be easy enough to do via a RwLock, or even without any synchronization, if properly initialized)
> 
> * The only difference between this and the previous case is that each runtime has its own challenge word, but use that for all incoming connections. Is that right?

Correct. This prevents a malicious Runtime instance from capturing the signed challenge from a user, and reusing it to authenticate for another Runtime as a MITM / replay attack.

> 
> > * per-connection challenge (needs more work to figure how to do it for both gRPC and HTTP)
> 
> * Each connection gets a different challenge. The same user may get different challenges for different connections. Is it correct?

Correct.

> 
> From the discussions it sounds like "per-runtime challenge" would cause the load-balancing problem, and the "per-connection challenge" approach can solve it. Does the "per-connection challenge" inherently solve this problem?! Or addressing this issue is what you mean by the "more work" that is needed for "per-connection challenge" @tiziano88 ?

"per-runtime challenge" would not cause the load-balancing problem, as long as the challenge is provided over the same (already established) gRPC / HTTP connection.

**Comment by conradgrobler** [comment]
> * The client produces a response (signed challenge and public key) and sends it to the server via another gRPC/HTTP call
> * All future calls on the connection will then have the privilege associated with the associated identity (this is from Conrad's description above, and I don't quite understand it. Can you explain @conradgrobler? Does the client have to sign all the messages or include the signature for the challenge in a header or something? What does the server do with the signature and public key that it receives from the client?)

Until we have per-connection state that can store the associated identity we cannot support a model where the client sends the signed challenge only once for all future calls on that connection, so for now the signed challenge must be included as a metadata in all future requests on that connection.

Regarding the load-balancing issue: If a client always starts a session on a specific connection with getting the challenge for that connection, load balancing should not be a problem. Load balancing would only present a problem if the client got the challenge on one connection and tries to reuse it on a different connection that connects to a different runtime.

**Comment by tiziano88** [comment]
> > * The client produces a response (signed challenge and public key) and sends it to the server via another gRPC/HTTP call
> > * All future calls on the connection will then have the privilege associated with the associated identity (this is from Conrad's description above, and I don't quite understand it. Can you explain @conradgrobler? Does the client have to sign all the messages or include the signature for the challenge in a header or something? What does the server do with the signature and public key that it receives from the client?)
> 
> Until we have per-connection state that can store the associated identity we cannot support a model where the client sends the signed challenge only once for all future calls on that connection, so for now the signed challenge must be included as a metadata in all future requests on that connection.

Ah you are right, it will be necessary until we have that state. But I think it is sufficient to send it once per **invocation**, rather (i.e. on the first request of a stream, in case of client streaming -- which we don't currently support anyways).

> 
> Regarding the load-balancing issue: If a client always starts a session on a specific connection with getting the challenge for that connection, load balancing should not be a problem. Load balancing would only present a problem if the client got the challenge on one connection and tries to reuse it on a different connection that connects to a different runtime.

This is my understanding too.

**Comment by rbehjati** [comment]
Thanks for the clarifications @tiziano88 and @conradgrobler. I still don't quite understand what the server (the Oak-Runtime) does with the responses or signed challenge in the metadata. Is it only used as IFC privileges and labels? I also don't see how the bearer token is currently used by the Oak-Runtime.

**Comment by conradgrobler** [comment]
Yes, it will only be used as the privilege to declassify messages with a user label containing the public key. I don't think this declassification logic is implemented for bearer tokens yet.

**Comment by tiziano88** [comment]
> Thanks for the clarifications @tiziano88 and @conradgrobler. I still don't quite understand what the server (the Oak-Runtime) does with the responses or signed challenge in the metadata. Is it only used as IFC privileges and labels? I also don't see how the bearer token is currently used by the Oak-Runtime.

You are right, the privilege / declassification logic is not actually implemented even for bearer tokens yet, I'll create a separate issue for it.

**Comment by tiziano88** [comment]
Created #1452

**Comment by rbehjati** [comment]
Thanks. Is that is in addition to the invocation channel labels that should be set based on the identity of the user (#1428)? I guess we should use the same authentication mechanism for that too.

**Comment by tiziano88** [comment]
> Thanks. Is that is in addition to the invocation channel labels that should be set based on the identity of the user (#1428)? I guess we should use the same authentication mechanism for that too.

Yes, the label on the invocation channels prevent data for other users from being sent on those channels, but we still need the gRPC node to only declassify data for the currently authenticated user.

**Comment by tiziano88** [comment]
Actually, to be precise, I think:

- the label on the request read channel should be the one specified via the oak label gRPC metadata
- the label on the response write channel should be the one determined via the challenge response authentication

then, the node would declassify the data from the response write channel before sending it back on the socket

**Comment by rbehjati** [comment]
I guess, according to the [docs](https://github.com/project-oak/oak/blob/main/docs/concepts.md#grpc-and-user-labels)), the integrity component of the label for the request receiver channel should as well be set to the user's identity via the challenge response authentication.

**Comment by conradgrobler** [comment]
I agree.

On a different point: although the response sender channel label should have an empty integrity component, should it not also be the same as the integrity component of the request receiver to support robust declassification? Otherwise, the gRPC pseudo node effectively endorses the integrity of the data by declassifying the data to the user without doing any checking or validation.

**Comment by rbehjati** [comment]
I think the current design does not support robust declassification. When we add the support for that, and remove the privileges, we have to change the labels accordingly.

**Comment by tiziano88** [comment]
@rbehjati is correct, currently the declassification is controlled by a separate privilege, which in the future may be merged with integrity label itself (and vice versa for endorsement).

The important thing to keep in mind is that , even in the case in which we have actual separate virtual nodes for each invocation, the check of whether a node may declassify data needs to be done voluntarily by the node itself. In other words, "native" (Rust) pseudo nodes are not sandboxed in any way, so we need to rely on manual inspection to ensure they are declassifying only data they have the privilege for.
